README.md
idp-engine is a reasoning engine for knowledge represented using the FO(.) language.
FO(.) (aka FO-dot) is First Order logic, with various extensions to make it more expressive:  types, equality, arithmetic, inductive definitions, aggregates, and intensional objects.
The idp-engine uses the Z3 SMT solver as a back-end.

It is developed by the Knowledge Representation group at KU Leuven in Leuven, Belgium, and made available under the [GNU LGPL v3 License](https://www.gnu.org/licenses/lgpl-3.0.txt).

See more information at [www.IDP-Z3.be](https://www.IDP-Z3.be).

Contributors (alphabetical):  Bram Aerts, Ingmar Dasseville, Jo Devriendt, Marc Denecker, Matthias Van der Hallen, Pierre Carbonnelle, Simon Vandevelde

# Installation

* Install [python3](https://www.python.org/downloads/) on your machine.
* Install [poetry](https://python-poetry.org/docs/#installation):
    * after that, logout and login if requested, to update `$PATH`
* Use git to clone https://gitlab.com/krr/IDP-Z3 to a directory on your machine, e.g., IDP-Z3
* Open a terminal in that IDP-Z3 directory
* If you have several versions of python3, and want to run on a particular one, e.g., 3.10:
    * run `poetry env use 3.10`
    * replace `python3` by `python3.10` in every command below
* Run `poetry install`
    * use `poetry install --no-dev` if you do not plan to contribute to IDP_Z3 development.


# Get started

To launch the Interactive Consultant web server:

* open a terminal in that directory and run `poetry run python3 main.py`
* open your browser at http://127.0.0.1:5000

The Web IDE is at http://127.0.0.1:5000/IDE


# Develop

You may want to read about the [technical documentation](http://docs.idp-z3.be/en/latest/code_reference.html) and the [Development and deployment guide](https://gitlab.com/krr/IDP-Z3/-/wikis/Development-and-deployment-guide).

Development on the web client requires additional installation:

* install [nvm](https://github.com/nvm-sh/nvm)
* `cd idp_web_client`
* `nvm install 11`
* `npm ci`

To launch the Interactive Consultant in development mode:

* open a terminal in the IDP-Z3 directory and run `poetry run python3 main.py` to launch the IDP-Z3 server
* `cd idp_web_client`
* `npm start`
* open your browser at http://127.0.0.1:4201


# Documentation

The user manual is in the `/docs` folder and can be locally generated as follows:
~~~~
poetry run sphinx-autobuild docs docs/_build/html
~~~~
To view it, open `http://127.0.0.1:8000`

The [documentation](https://docs.IDP-Z3.be) on [readthedocs](https://readthedocs.org/projects/idp-z3/) is automatically updated from the main branch of the GitLab repository.

The [home page](https://www.IDP-Z3.be) is in the `/homepage` folder and can be locally generated as follows:
~~~~
poetry run sphinx-autobuild homepage homepage/_build/html
~~~~
To view it, open `http://127.0.0.1:8000`.  The website is also automatically updated from the main branch of the GitLab repository.


# Testing

To generate the tests, from the top directory run `poetry run python3 test.py` or `poetry run python3 test.py generate`.
After this, you can manually check what has changed using git.

There is also a testing pipeline available, which can be used by running `poetry run python3 test.py pipeline`.


# Testing of the Interactive Consultant

We use [Cypress](https://www.cypress.io/) to test the IC.  Installation in top project folder:
~~~~
npm install cypress --save-dev
~~~~
To launch it, start the Interactive Consultant in development mode, then:
~~~~
npx cypress open
~~~~
and choose E2E testing.

# Deploy

See the instructions [here](https://gitlab.com/krr/IDP-Z3/-/wikis/Development-and-deployment-guide).


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).
This project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html) relative to the user interface and to the API exposed by the web server.

## [0.11.1] - unreleased yet

The main feature of this release is type checking.

Breaking change: some KB may not pass the type check anymore,
in particular those declaring numeric types without interpretation (e.g. `type row`).
See `Explanations / Type checking` in the documentation for details.

Contributors (alphabetical): Pierre Carbonnelle.

### IDP-Z3
* MR !336: Improve type checking
* issue #309: "'AQuantification' object has no attribute 'symbol'" error in IC


### Interactive Consultant and Web IDE


## [0.10.12] - 2023-11-13

The main feature of this release is performance improvements.

Contributors (alphabetical): Pierre Carbonnelle, Simon Vandevelde.

### IDP-Z3
* MR !329: support full propagation of function values
* issue #289: Unsatisfiable theory without explanation
* issue #293: add stable semantics + fix completion semantics
* issue #294: speed up transitive closure of graph (+ fix import of deepcopy)
* issue #296: improve performance of theories with large enumerations
* issue #299: Order of declarations matter for symbols in definitions
* issue #302: avoid "maximum recursion depth exceeded" error for large enumerations


### Interactive Consultant and Web IDE
* issue #301: Propagation error in IC for simple theory
*
*


## [0.10.11] - 2023-08-18

The main feature of this release is performance improvements.

Contributors (alphabetical): Pierre Carbonnelle, Simon Vandevelde.

### IDP-Z3
* MR !319: improve some error messages
* MR !321: performance improvements in the grounding of quantifications
* MR !325: reimplement propagation using Z3's consequences
* MR !326: performance improvements in model expansion
* issue #3: fix issue with out of bounds numbers
* issue #32: added benchmark in MR !322
* issue #38: typehints throughout code
* issue #41: various quick syntax changes
* issue #46: use flake8 linter
* issue #67: reworked sliders in annotations
* issue #129: 0 == 0.0 is now correct
* issue #267: Use variable declaration for type inference in recursive definitions
* issue #284: incorrect propagation when quantifying over infinite domain
* issue #285: Develop benchmark suite


### Interactive Consultant and Web IDE
* issue #248: detect error for incorrect entry in a Int range
* issue #283: Allow loading a default theory when theory is incorrect


## [0.10.9] - 2023-05-31

This release has new package dependencies: please run `poetry install` to update.

Contributors (alphabetical): Pierre Carbonnelle, Simon Vandevelde.

### IDP-Z3
* MR !312, !315: refactor `simplify.py`
* MR !317: accelerate generation of type constraints
* fix #277: Incorrect relevance for theories with numeric comparison
* fix #279: Incorrect relevance for Registration

### Interactive Consultant and Web IDE
* issue #274: Content of editor is lost on page reload
* issue #278: Annotations in IDE: prettify bug and syntax highlight


## [0.10.8] - 2023-05-05

The main new features in this release are:
* new syntax for aggregates (MR !294)

Contributors (alphabetical): Pierre Carbonnelle, Simon Vandevelde.

### IDP-Z3
* MR !304: show line number for internal error in python code
* MR !305: IEP 27 `if` in sum / min / max aggregates
* issue #271: Incomplete interpretation of set of integers (+ MR !310)

### Interactive Consultant and Web IDE
* MR !302: incorrect linter error for partial interpretation of function
* issue #268: Open FLAIR link in new tab
* issue #273: Access to documentation is broken


## [0.10.4] - 2023-04-17

The main new features in this release are:
* support for recursive data types (#185)

This release has new package dependencies: please run `poetry install` to update.

Contributors (alphabetical): Pierre Carbonnelle, Simon Vandevelde.

### IDP-Z3
* MR !294: `are necessary and sufficient conditions for`
* issue #185: support for recursive datatypes
* issue #249: regression of performance with new Z3 solver
* issue #254: Improve type checking on inequalities
* issue #255: Line numbers are wrong
* issue #257: Error messages for uninterpreted types
* issue #262: "ground" in error messages

### Interactive Consultant and Web IDE
* MR !300: improve loading time in IC
* MR !301: access proper version of documentation
* issue #252: incorrect relevance with if-then-else
* issue #258: Disable (certain) style warnings


## [0.10.3] - 2023-03-24

The main new features in this release are:
* Support for co-inductive definitions (MR !279)
* connectives in English (MR !290)
* automated testing of the IC using Cypress  (See README.md for installation) (#242)

This release has new package dependencies: please run `poetry install` to update.

Contributors (alphabetical): Pierre Carbonnelle, Simon Vandevelde.

### IDP-Z3
* MR !279: Support co-inductive definitions
* MR !280: various small improvements
* MR !284: Refactor Expression.simpler for better performance
* MR !290: English connectives in formula
* issue #235: range interpretation of predicate is not respected
* issue #236: Incorrect nested quantification over infinite domain
* issue #238: Fix FOLint issues in structure
* issue #240: internal error with quantification over infinite domain
* issue #246: error with chained inequalities

### Interactive Consultant and Web IDE
* MR !282: improve display of ⇒ in the editor.
* MR !289: use narrower numeric field
* issue #192: access to gist temporarily disabled
* issue #241: Environmental verification error in the Covid example
* issue #245: Fix bugs in the Save functionality
* issue #242: Can't retract assertion in Polygon (+ Cypress)
* issue #247: issues with Date field


## [0.10.2] - 2022-12-19

Contributors (alphabetical): Pierre Carbonnelle.

### IDP-Z3

deploy.py
"""
Script to deploy the IDP-Z3 project, to be used whenever there is a new version
ready.

To run it:
    * commit your last changes in the repository
    * `poetry env use 3.11 & poetry update`
    * start the web and client servers
    * `poetry run python3.11 deploy.py`

It does the following things:
    * add a tag in git;
    * update version in pyproject.toml;
    * deploy idp-engine package to pypi;
    * deploy folint package to pypi;
    * deploy IDP-Z3 to GAE;
    * add version in versions.json.

Before actually deploying, it runs the tests to make sure that there are no
errors, and regenerates the '/IDP-Z3/idp_web_server/static' folder.

Authors: Pierre Carbonnelle, Simon Vandevelde
"""
from distutils.dir_util import copy_tree
import json
import re
import requests
import subprocess
import sys


def run(c, **kwargs):
    return subprocess.run(c.split(' '), **kwargs)


def get(c, **kwargs):
    return run(c, capture_output=True, **kwargs).stdout


def require_clean_work_tree(cwd):
    assert 0 == run('git diff-files --quiet --ignore-submodules --', cwd=cwd).returncode, \
        "Cannot deploy: you have unstaged changes."
    assert 0 == run('git diff-index --cached --quiet HEAD --ignore-submodules --', cwd=cwd).returncode, \
        "Cannot deploy: your index contains uncommitted changes."
    run('git pull', cwd=cwd)  # may fail if there are commits


def query_user(query, default="y", get=False):
    if get is True:
        return input(query)
    if default == "y":
        return input(query) in "Yy"
    else:
        return input(query) in "Nn"

def running(url):
    try:
        response = requests.get(url)
        return response.status_code == 200
    except requests.exceptions.RequestException as e:
        return False

if not query_user("Did you do a 'poetry update'? (Y/n)"):
    sys.exit()

# run Cypress tests
if not running("http://localhost:5000"):
    query_user("Please start web server.")
if not running("http://localhost:4201"):
    query_user("Please start web client.")
run("npx cypress open")

run('git pull origin', check=True)
run('python3 test.py generate')

if query_user("Run benchmark? (Y/n)"):
    run('python3 test.py benchmark')

update_statics = query_user("Update the '/IDP-Z3/idp_web_server/static' folder? (Y/n) ")
if update_statics:

    # Generate static and commit.
    run('npm run-script build', cwd='idp_web_client', check=True)
    print("Copying to static folder ...")
    copy_tree('idp_web_client/dist/', './idp_web_server/static')

    # Create new version tag.
    new_tag = query_user("Create new tag? (Y/n) ")
    if new_tag:
        # Verify we are on main branch.
        branch = get('git rev-parse --abbrev-ref HEAD')
        assert branch == b'main\n', \
            "Cannot deploy: IDP-Z3 not in main branch !"

        # Find old tag.
        current_tag = (get("git describe --tags --abbrev=0")
                       .decode("utf-8").strip())  # Strip newline
        major, minor, patch = current_tag.split('.')
        release_type = query_user("(M)ajor, (m)inor or (p)atch release? ",
                                  get=True)
        # Compute new tag
        if release_type == "M":
            tag_version = f"{int(major)+1}.0.0"
        elif release_type == "m":
            tag_version = f"{major}.{int(minor)+1}.0"
        elif release_type == "p":
            tag_version = f"{major}.{minor}.{int(patch)+1}"
        else:
            raise IOError("Incorrect release type")
        assert query_user(f"Release {tag_version} ? (Y/N)")

        # We also need to modify the pyproject.toml.
        with open("./pyproject.toml", "r") as fp:
            pyproject = fp.read()
        pyproject = re.sub(r'version = ".*"',
                           f'version = "{tag_version}"',
                           pyproject)
        with open("./pyproject.toml", "w") as fp:
            fp.write(pyproject)

        _ = query_user("Update version (and contributors) in CHANGELOG.  Ready ?(Y/N)")
        _ = query_user("Update version in IDP-Z3.py.  Ready ?(Y/N)")
        _ = query_user("Update folint and IDP-Z3 version in /folint/setup.py. Ready ?(Y/N)")

    # Add and commit.
    run("git add -A")
    run("git commit")
    run("git push origin main")

    if new_tag:
        # create tag
        run(f"git tag {tag_version}")

        # Push tags.
        run(f"git push origin {tag_version}")

        # Publish new version of IDP-Z3 on Pypi.
        run("poetry install")
        run("poetry build")
        print("Publishing to pypi (Username: krr)")
        run("poetry publish") #  use `poetry config http-basic.pypi <username> <password>` to set up
        run("rm -rf ./dist")

        # publish new version of folint on Pypi
        run("python3.11 setup.py sdist bdist_wheel", cwd="folint")
        run("poetry run python3.11 -m twine upload dist/*", cwd="folint")

    # if input("Deploy on Heroku ?") in "Yy":
    #     run("git push heroku main")

    if new_tag or query_user("Deploy to Google App Engine? (Y/n) "):
        print("Deploying to GAE")

        # Push to google repository
        # run("git push google main")
        # run("git push google main", cwd='idp_web_client')

        # deploy to GAE
        run(f"gcloud app deploy {'' if new_tag else '--no-promote'}")

        # update versions.list at https://gist.github.com/IDP-Z3/5d82c61fa39e8aa23da1642a2e2b420a
        versions = get("gcloud app versions list --sort-by=~LAST_DEPLOYED").decode("utf-8")
        id = versions.splitlines()[1].split("  ")[1]
        print("latest version : ", id)
        with open("../5d82c61fa39e8aa23da1642a2e2b420a/versions.json") as json_file:
            data = json.load(json_file)
        data['IDP-Z3 latest'] = f"{id}-dot-interactive-consultant.ew.r.appspot.com"
        if new_tag:
            data[f'IDP-Z3 {tag_version}'] = data['IDP-Z3 latest']
        with open("../5d82c61fa39e8aa23da1642a2e2b420a/versions.json", "w") as outfile:
            json.dump(data, outfile, indent=4)
        run("git add versions.json" , cwd="../5d82c61fa39e8aa23da1642a2e2b420a")
        run('git commit -m "latest"', cwd="../5d82c61fa39e8aa23da1642a2e2b420a")
        run("git push origin master", cwd="../5d82c61fa39e8aa23da1642a2e2b420a")

        # open browser on GAE
        run(f"browse https://{id}-dot-interactive-consultant.ew.r.appspot.com/")
